File: implementation/src/main/java/ro/anud/xml_xsd/implementation/util/LinkedNode.java
package ro.anud.xml_xsd.implementation.util;

import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public interface LinkedNode {


    String classTypeId();

    Optional<LinkedNode> parentNode();
    RawNode serializeIntoRawNode();
    void removeChild(Object object);

    default String nodeName() {
        return "";
    }

    RawNode rawNode();

    void notifyChange(LinkedNode linkedNode);
    default void notifyChange() {
        notifyChange(this);
    }
    void notifyRemove(LinkedNode linkedNode);
    default void notifyRemove() {
        notifyRemove(this);
    }

    default void clearParentNode() {
        notifyRemove();
        var parentNode = parentNode();
        parentNode = Optional.empty();
        parentNode.ifPresent(LinkedNode::notifyChange);
    }

    int buildIndexForChild(Object object);

    LinkedNode deserializeAtPath(String xpath, RawNode rawNode);

    default String buildPath() {
        var index = parentNode().map(linkedNode -> linkedNode.buildIndexForChild(this)).orElse(0);
        var path = "." + nodeName() + "[" + index + "]";

        return parentNode()
            .map(LinkedNode::buildPath)
            .map(string -> string + path)
            .orElse(path);
    }
}

File: gui-client/xsd/LinkedNode.cs
using System;
using System.Collections.Generic;
using System.Xml;

namespace XSD {
    public interface ILinkedNode
    {
        public ILinkedNode? ParentNode { get; set; }
        
        public string NodeName { get;}
        public int? BuildIndexForChild(ILinkedNode linkedNode);
        
        void NotifyChange(List<ILinkedNode> linkedNodes);

        public RawNode SerializeIntoRawNode();
        
        public bool IsValidChildType(ILinkedNode candidateChild);

        public void Serialize(XmlElement element);
        
        public void SetAttribute(string name, string? value);
        
        public void SetChild(dynamic linkedNode);
        public void ClearChild(dynamic linkedNode);

        public void DeserializeAtPath(string xpath, RawNode rawNode);

        public Action OnChange(Action<List<ILinkedNode>> callback);
        public Action OnSelfChangeNode(Action<ILinkedNode> callback);

        public string BuildPath()
        {
            var index = ParentNode?.BuildIndexForChild(this) ?? 0;
            var path = "." + NodeName + "[" + index + "]";
            if (ParentNode == null)
            {
                return path;
            }
            return ParentNode.BuildPath() + path;
        }

    }
}